### PR TITLE
feat: Added Warshall's Transitive Closure to Graph folder

### LINKED ISSUE #2063: Add Warshall's Transitive Closure Algorithm to Graph folder
### Detailed description

Warshall's algorithm is used to determine the transitive closure of a directed graph or all paths in a directed graph by using the adjacency matrix. For this, it generates a sequence of n matrices. Where, n is used to describe the number of vertices.

### Context

In mathematics, the transitive closure of a binary relation R on a set X is the smallest relation on X that contains R and is transitive. In computer science, the concept of transitive closure can be thought of as constructing a data structure that makes it possible to answer reachability questions.  Warshall's algorithm can be implemented in both cases. 

### Possible implementation

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by abhilipsasahoo03**
#### Description of Change

- Added Warshall's Transitive Closure in C++ to Graph folder
- Closes #2063

#### Screenshot

Here is a screenshot of the output
Screenshot_2022-10-02-22-35-00-213_com kvassyu coding2 cpp

You can check if it's correct or not by visiting the example [here](https://www.javatpoint.com/closure-properties-of-relations#:~:text=The%20transitive%20closure%20R%20of,smallest%20transitive%20relation%20containing%20R.&text=Suppose%20A%20is%20a%20finite%20set%20with%20n%20elements.&text=Example2%3A%20Let%20A%20%3D%20%7B4,a%20relation%20on%20set%20A.)

#### Checklist

- [X] Added description of change
- [X] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [X] Relevant documentation/comments is changed or added
- [X] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [X] Search previous suggestions before making a new one, as yours may be a duplicate.
- [X] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

**Review by mishraabhinn — CHANGES_REQUESTED** [review]
Have you read [Contributing Guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md)
If not I really emphasize reading this. 
Thanks. 😃

**Comment by abhilipsasahoo03** [comment]
Hi @mishraabhinn I've updated documentation with a sample output. PTAL.

File: graph/warshall_transitive_closure.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+The resulting matrix after following the steps n times and obtaining
+n number of matrices, the nth matrix is the transitive closure for
+the given adjacency matrix.
+
+@author [Abhilipsa Sahoo](https://github.com/abhilipsasahoo03)
+*/
+
+#include<iostream> //For IO Operations
+using namespace std;
```
Comment (mishraabhinn) [inline graph/warshall_transitive_closure.cpp:41]: remove this

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

File: graph/warshall_transitive_closure.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        }
+        else if(ans == 1)
+        {
+          Arr[i-1][j-1] = 1;
+        }
+      }
+    }
+    cout<<"\nAdjacency matrix is: \n";
+    Print_Matrix(Arr, n);
+    Transitive_Closure(Arr, n);
```
Comment (mishraabhinn) [inline graph/warshall_transitive_closure.cpp:70]: Need some automatic tests for this. Can you please do that ?
```
static void test() {
    /* descriptions of the following test */
    assert(func(...) == ...); // this ensures that the algorithm works as expected

    // can have multiple checks
}

```

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

**Comment by abhilipsasahoo03** [comment]
Hi @mishraabhinn I've added the tests and made required changes as suggested by you. PTAL.

**Comment by abhilipsasahoo03** [comment]
@mishraabhinn @Panquesito7 kindly review the changes that have been made and please do let me know if anything else is required.

File: graph/warshall_transitive_closure.cpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  for(k=0; k<node; k++)
+  {
+    for(row=0; row<node; row++) 
+    {
+      for(col=0; col<node; col++) 
+      {
+        if((Arr[row][col]==1)||(Arr[row][k]==1 && Arr[k][col]==1))
+         {
+            Res[row][col] = 1;
+         }
```
Comment (mishraabhinn) [inline graph/warshall_transitive_closure.cpp:115]:
```suggestion
      Res[i][j] = Arr[i][j] ||  (Arr[i][k] && Arr[k][j]);
```

How about this? Will this work ?

**Review by mishraabhinn — CHANGES_REQUESTED** [review]

File: graph/warshall_transitive_closure.cpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+int Transitive_Closure(int Arr[10][10], int node);
+void Print_Matrix(int A[10][10], int n);
+static void test();
+int main()
+{
+    cout<<"\nInitializing tests...";
+    test();
+    int n, i, j, ans;
+    int Arr[10][10];
```
[the remark this answers is not in the document]
Comment (abhilipsasahoo03) [reply graph/warshall_transitive_closure.cpp:52]: In Warshall's algo, Θ(n3) is the worst-case cost. Even Brute Force method seems a viable option when compared to this. This is why I tried to implement a static 2D array with fixed size of 10 inputs to maintain responsiveness.

File: graph/warshall_transitive_closure.cpp (hunk 3 above)
Comment (abhilipsasahoo03) [reply graph/warshall_transitive_closure.cpp:115]: I tried to work this in an online compiler, and the adjacency matrix is simply copied to the transitive closure with this logic, which is not the desirable output. 

Sample screenshot:

File: graph/warshall_transitive_closure.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (abhilipsasahoo03) [reply graph/warshall_transitive_closure.cpp]: As I focussed on static implementation, I thought of static Array as a much suited data structure in this scenario and didn't think of vector as a necessity.

**Comment by abhilipsasahoo03** [comment]
Hi @mishraabhinn I've responded to your requested changes. Kindly take a look.

**Review by mishraabhinn — APPROVED** [review]
Cool. Nice work @abhilipsasahoo03. I am approving this PR. Looking forward to some more contributions. Thanks. You can join Gitter and discord for more connected to TheAlgorithms.

**Comment by abhilipsasahoo03** [comment]
> Cool. Nice work @abhilipsasahoo03. I am approving this PR. Looking forward to some more contributions. Thanks. You can join Gitter and discord for more connected to TheAlgorithms. 

Thank you so much! Your feedback was insightful!

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Hey there! Thank you for submitting a PR! 😄 Sadly, the documentation is not up to our [**standards**](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md). Please update the file accordingly. **Let us know if you need any help here or on our [Discord server](https://the-algorithms.com/discord).** 😃

**Comment by abhilipsasahoo03** [comment]
Hi @Panquesito7 I've made some changes to the documentation. Could you please review it and let me know where and what exactly do I need to improve further if needed?

File: graph/warshall_transitive_closure.cpp [hunk 5]
```diff
@@ -0,0 +1,179 @@
+/**
+* @file
+* @brief Implementation of Warshall's Algorithm in C++
+* [Warshall's Transitive Closure Algorithm](https://www.javatpoint.com/warshalls-algorithm)
```
Comment (Panquesito7) [inline graph/warshall_transitive_closure.cpp:4]:
```suggestion
* @brief Implementation of [Warshall's Algorithm](https://www.javatpoint.com/warshalls-algorithm) in C++
```

File: graph/warshall_transitive_closure.cpp [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+* The resulting matrix after following the steps n times and obtaining
+* n number of matrices, the nth matrix is the transitive closure for
+* the given adjacency matrix.
+*
+* @author [Abhilipsa Sahoo](https://github.com/abhilipsasahoo03)
+*/
+
+
+#include<iostream> //For IO Operations
+#include<cassert> //For using assert()
```
Comment (Panquesito7) [inline graph/warshall_transitive_closure.cpp:50]:
```suggestion
#include<iostream> /// for IO Operations
#include<cassert> ///  for assert
```

File: graph/warshall_transitive_closure.cpp [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+
+/** Automating tests to check if the algorithm runs perfectlyperfectly
+* for two test cases
+* @return None
+*/
+static void test()
```
Comment (Panquesito7) [inline graph/warshall_transitive_closure.cpp:57]: The tests should go after all the functions and before `main`.

File: graph/warshall_transitive_closure.cpp [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+{
+  //Test 1
+  int Arr1[10][10] = {{1, 0, 0, 1}, {0, 1, 1, 0}, {0, 0, 0, 1}, {0, 0, 0, 0}};
```
Comment (Panquesito7) [inline graph/warshall_transitive_closure.cpp:60]: Due to the use of static/C-style arrays, the CI is specifying we should use `std::array` or `std::vector` instead.
This should be still implementable and should work fine with this algorithm. If you need any help, let us know. 🙂

File: graph/warshall_transitive_closure.cpp [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+  assert(Transitive_Closure(Arr2, 4) == 11565);
+  cout<<"\nTest 2 completed";
+  cout<<"\nAll tests completed successfully";
+
+}
+
+/** Finds Transitive Closure for the adjacency matrix
+* and calculates and assigns a unique value to return
+* at the time of assertion
+* @param integer type Adjacency matrix Arr of nxn where
```
Comment (Panquesito7) [inline graph/warshall_transitive_closure.cpp:75]: Both parameters should be documented _separately_ (example).
```c++
/**
 * @brief This is my awesome function
 * @details
 * Detailed explanation of my awesome function
 *
 * @param my_param This parameter is used for...
 * @param my_param2 This parameter is used for...
 * @param my_array This array is used for the algorithm
 * @returns the 2nd index of  `my_array`
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 🚀

**Pull request closed without merging**